![example branch parameter](https://github.com/jlab/fold-grammars/actions/workflows/c-cpp.yml/badge.svg)

# fold-grammars
Collection of bgap code for RNA folding
For rudimentary documentation, consult https://bibiserv.cebitec.uni-bielefeld.de/fold-grammars/

# RNALoops

RNALoops makes "motified" algebras available, allowing for the incorporation of RNA 3D Motifs into RNA secondary structure predictions. There are currently four algebras available: ```alg_motif```, ```alg_pretty```, ```alg_shapes_mot``` and ```alg_hishapes_mot```, findable under ```RNALoops/Algebras/Motif```. ```Motif```, ```shapes``` and ```hishapes``` can be used as classifying algebras in the Bellman's GAP system. ```pretty``` is a adapted DotBracket Algebra that incorporates the found motifs into a DotBracket depiction. The underlying C++ implementation for the motifs can be found in the ```motif.hh``` header file under ```RNALoops/Extensions```.
RNALoops is currently only available for UNIX systems (everything was written and tested on Ubuntu 22.04.3 LTS and python 3.10.12 64-bit).

Setup:</br>
    1. Install the gapcM compiler from ```https://github.com/RNABioInfo/gapcM.git```, which is a slightly altered gapc compiler. It retains full functionality and has the exact same set up as the original (I keep the repo up to date with the original gapc from ```https://github.com/jlab/gapc.git```). The changes made to the original are necessary for motif implementations.</br>
    2. Now you can run the ```RNALoops.py``` python script (located in the main ```RNALoops``` folder) with several preset algorithms: ```motmfepretty```, ```motshapeX```, ```mothishapes```, ```motpfc```, ```motshapeX_pfc```, ```mothishapes_h_pfc```, ```mothishapes_b_pfc```, ```mothishapes_m_pfc```. All non partition function algorithms can also be used with suboptimal folding by adding -s as a command line argument. For information on how to structure your inputs, help can be called for with ```RNALoops.py -h```. Additionally, you do not need to compile all the algorithms yourself with the GAP Compiler, the Python script automatically checks if a chosen algorithm is already compiled and compiles it if it isn't there. If you want to use the python script for non preset algorithms, you can add a new instance to ```RNALoops.gap``` or pre-compile it yourself. If you do not precompile you have to add a new call construct to the Process.call_constructor function or add you algorithm's name to one of the existing cases, depending on which arguments you want to set for calling it. If you want use a partition function algorithm make sure to end the name in pfc, as this tag signals the script to calculate class probabilities from the partition function values. Motif sequence collection has been update to be automatically run if used version is not the up to date one (this can be turned off). Motif sequences are saved as hexadecimals in C style include arrays, saved in the header file ```mot_header.hh``` in the extensions subfolder.  To specify which motifs to use change the ```motifs.json``` file in ```src/data```. This allows you to add or remove motifs from the generated C arrays so you can customize your predictions as you please. For this python script you might need to install AlignIO from Bio, if it is not already part of your python environment. 
    The TSV output currently has three different output formats: ```ID,class1,score,class2``` |```ID,class,score ```| ```ID,class,score,probability```. The last case is in particular used with all partition function implementations to directly calculate probabilities for each output class. Which header is used is determined by the script automatically.</br>
    ```RNALoops.py```  allows you to input fasta, fastq or stockholm files and it will run your specified algorithm on all sequences in the file. Outputs are formatted to csv and piped to stdout. Which separation character is to be used can be customized, default is a tab. Logs are piped to stderr. With the -t argument you can additionally run each prediction with the ```time``` command, giving the option to monitor runtimes and memory usage for predictions. If the automated algorithm compilation functionality does not work, setting the loglevel to debug will output errors from the compilation. Algorithms are automatically compiled into the ```RNALoops``` base folder. If you want to use a log file to avoid long command line calls, -f allows for a config file to be used (this is currently set up in ```src/data``` as even without -f the file is used to keep tabs on which motif sequence version you are currently on, please do not move or remove this file.)

If anything should not work for you when trying to implement RNALoops, please feel free to reach out to me through my public e-mail.
